Src/Support/Google.Apis.Auth/OAuth2/AwsExternalAccountCredential.cs

```diff
@@ -0,0 +1,168 @@
+                SecurityCredentialsUrl = other.SecurityCredentialsUrl;
+                ImdsV2SessionTokenUrl = other.ImdsV2SessionTokenUrl;
+            }
+
+            internal Initializer(AwsExternalAccountCredential other) : base(other)
+            {
+                RegionUrl = other.RegionUrl;
+                RegionalCredentialVerificationUrl = other.RegionalCredentialVerificationUrl;
+                SecurityCredentialsUrl = other.SecurityCredentialsUrl;
+                ImdsV2SessionTokenUrl = other.ImdsV2SessionTokenUrl;
+            }
+        }
+
+        /// <summary>
+        /// Metadata server URL used to obtained the region that should be included as part of
+        /// the subject token.
+        /// </summary>
+        internal string RegionUrl { get; }
+
+        /// <summary>
+        /// STS server will use this URL to validate the subject token included
+        /// on the STS request. This URL will be included as part of the subject token.
+        /// </summary>
+        internal string RegionalCredentialVerificationUrl { get; }
+
+        /// <summary>
+        /// Metadata server URL from which to obtain the security credentials that will
+        /// be used to sign the subject token.
+        /// </summary>
+        internal string SecurityCredentialsUrl { get; }
+
+        /// <summary>
+        /// If present, a session token fetched from this URL should be used when making
+        /// requests to the metadata server.
+        /// </summary>
+        internal string ImdsV2SessionTokenUrl { get; }
+
+        internal AwsExternalAccountCredential(Initializer initializer) : base(initializer)
+        {
+            RegionUrl = initializer.RegionUrl;
+            RegionalCredentialVerificationUrl = initializer.RegionalCredentialVerificationUrl;
+            SecurityCredentialsUrl = initializer.SecurityCredentialsUrl;
+            ImdsV2SessionTokenUrl = initializer.ImdsV2SessionTokenUrl;
+        }
+
+        /// <inheritdoc/>
+        private protected override GoogleCredential WithoutImpersonationConfigurationImpl() =>
+            ServiceAccountImpersonationUrl is null
+            ? new GoogleCredential(this)
+            : new GoogleCredential(new AwsExternalAccountCredential(new Initializer(this)
+            {
+                ServiceAccountImpersonationUrl = null
+            }));
+
+        /// <inheritdoc/>
+        protected async override Task<string> GetSubjectTokenAsyncImpl(CancellationToken taskCancellationToken)
+        {
+            // We might not need to use the metadata server at all,
+            // but initializing this client here triggers no requests, etc.
+            // and makes the code cleaner.
+            var metadataClient = new AwsMetadataServerClient(ImdsV2SessionTokenUrl, HttpClient);
+
+            var awsRegion = await AwsRegion.FetchAsync(metadataClient, RegionUrl).ConfigureAwait(false);
```

Comment:
What if fetching the session token fails though, and it wouldn't have been used if a metadata server call is not used (e.g. due to environment variables)? I strongly suspect your approach is *better*, but it may be worth clarifying whether it's actually "correct" as per the spec. (We might want to change the AIP.)

Reply:
Well, since the token is only used if the metadata server is needed, I'm not sure we should fail in the case that the env vars are present but fetching the token fails. But yes, I see your point now and in contrast with the AIP. I'll add a couple more commits to the AIP about this, to make it reflect my implementation.